browser/ui/tabs/brave_tab_strip_model.h

```diff
@@ -0,0 +1,61 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BROWSER_UI_TABS_BRAVE_TAB_STRIP_MODEL_H_
+#define BRAVE_BROWSER_UI_TABS_BRAVE_TAB_STRIP_MODEL_H_
+
+#include <memory>
+#include <vector>
+
+#include "chrome/browser/ui/tabs/tab_strip_model.h"
+
+
+#include "ui/events/event_handler.h"
+
+
+class BraveTabStripModel : public TabStripModel {
+ public:
+  explicit BraveTabStripModel(TabStripModelDelegate* delegate,
+                               Profile* profile);
+
+  ~BraveTabStripModel() override;
+
+  void SelectNextTab(UserGestureDetails detail =
+    UserGestureDetails(GestureType::kOther)) override;
+
+  void SelectPreviousTab(UserGestureDetails detail =
+    UserGestureDetails(GestureType::kOther)) override;
+
+  // Set the next tab when doing a MRU cycling with Ctrl-tab
+  void SelectTabMRU(bool backward,
+  UserGestureDetails detail = UserGestureDetails(GestureType::kOther));
+
+  // Stop MRU cycling, called when releasing the Ctrl key
+  void StopMRUCycling();
+
+ private:
+  // To capture the release of the Ctrl key when doing a MRU cycling with
+  // Ctrl-tab
+  class CtrlReleaseHandler : public ui::EventHandler {
+   public:
+    explicit CtrlReleaseHandler(BraveTabStripModel* tab_strip);
+    ~CtrlReleaseHandler() override;
+
+   private:
+    void OnKeyEvent(ui::KeyEvent* event) override;
+    BraveTabStripModel* tab_strip;
+    DISALLOW_COPY_AND_ASSIGN(CtrlReleaseHandler);
+  };
+
+  std::unique_ptr<CtrlReleaseHandler> ctrl_released_event_handler;
+
+  // Current index when we are MRU cycling, set to -1 when not cycling
+  int current_mru_cycling_index = -1;
+
+  // List of tab indexes sorted by most recently used
+  std::vector<int> mru_cycle_list;
```

Comment:
have you checked `base::MRUCache`?

Reply:
I was not aware of this class but I am not sure it is a good fit here. I am calculating from scratch the ordered set of tab indices at the beginning of each cycling session with `std:sort` and the timestamp of last use.
`base:MRUCache` keeps track of the ordering by calling the `Put` function whenever the item is used.

To use `MRUCache` I would need to call `Put` each time a tab is accessed, not sure this is possible form `TabStripModel`.

Reply:
Hmm. So it looks like that timestamp is being set in `TabStripModel::ActivateTabAt`.  What about overriding that method and calling `Put` when the new tab is activated?

Reply:
My problem with `base::MRUCache` is that it's a map data-structure with key-value pairs. What I need is a simple MRU queue.
Also, I thought that relying on this already existing timestamp would be better than maintaining an additional index.